Add support for verifying root checksum in cosign initialize

bkabrda opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficient time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary

The motivation for this PR is explained in and closes https://github.com/sigstore/cosign/issues/3946 - the ability to verify initial root checksum when downloading via a URL.

Some notes:
* The `DoInitialize` function still works exactly the same; it's unused here, but may be used from elsewhere (like gitsign).
* Once this PR is merged, I will send another one to utilize the new functionality in gitsign as well.
* I wasn't sure what deprecation date to set here, so I'd be glad for advice.

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.
-->

Add a new `--root-checksum` flag for `cosign-initialize`. This will ensure cosign verifies the checksum of the file provided via the `--root` argument before proceeding with initialization. Using this argument will be mandated in the future for initial root obtained from URL.

#### Documentation

I don't think this requires specific callout in docs, but do let me know if you think oterwise.

haydentherapper commented on cmd/cosign/cli/options/deprecate.go:
```diff
@@ -19,3 +19,8 @@ const SBOMAttachmentDeprecation = "WARNING: SBOM attachments are deprecated " +
+
+const RootWithoutChecksumDeprecation = "WARNING: Fetching initial root from URL " +
+	"without providing its checksum is deprecated and will be disallowed in " +
+	"a Cosing release soon after TODO. Please provide the initial root checksum " +
```
```suggestion
	"a future Cosign release. Please provide the initial root checksum " +
```

haydentherapper commented on cmd/cosign/cli/options/initialize.go:
```diff
@@ -36,4 +37,7 @@ func (o *InitializeOptions) AddFlags(cmd *cobra.Command) {
 	_ = cmd.Flags().SetAnnotation("root", cobra.BashCompSubdirsInDir, []string{})
+
+	cmd.Flags().StringVar(&o.RootChecksum, "root-checksum", "",
+		"checksum of the initial root, required if root is downloaded via http(s). expects sha512 by default, can be changed to sha256 by providing sha256:<checksum>")
```
Can we flip this and have sha256 as the default? sha512 is probably overkill.

bkabrda replied:
Sure, will do.

haydentherapper reviewed (commented):
Thanks!

haydentherapper commented:
Just bumping!

bkabrda commented:
@haydentherapper I addressed your review - thanks for providing it! Sorry for the delay, I had a long end of year break where I was not at a computer.
